File: leetcode/reverse/reverse.py
# Given a 32-bit `signed integer, reverse digits of an integer.

# Example 1:
# Input: 123
# Output: 321

# Example 2:
# Input: -123
# Output: -321

# Example 3:
# Input: 120
# Output: 21

# !!!!!Time Limit Exceeded
# class Solution:
#     def reverse(self, x):
#         """
#         :type x: int
#         :rtype: int
#         """
#         e=1
#         temp = x
#         while True:
#             temp = temp // 10
#             if(temp):
#                 e += 1
#             else: 
#                 break
#         result = 0
#         for i in range(e):
#             mod = x % 10
#             x = x // 10
#             a = 10**(e-1-i) * mod
#             result += a
#         return result

class Solution: 
    def reverse(self, x): 
        """ 
        :type x: int 
        :rtype: int 
        """ 
        if x<0: 
            sign = -1 
        else: 
            sign = 1 
        strx=str(abs(x)) 
        r = int(strx[::-1]) 
        if r > 2147483647: 
            return 0; 
        return sign*int(r)
    def reverse_v2(self, x):
        res = 0
        if x > 2147483647: 
            return 0; 
        while x != 0:
            if(abs(res)> 2147483647 / 10):
                return 0
            res = res * 10 + x % 10
            x = x // 10
        return res
        

if __name__ == '__main__':
    solver = Solution()
    print(solver.reverse_v2(1000000009))
    # print(solver.reverse(120))
    # print(solver.reverse(1))
    # s = list(str(123))
    # print(s)
    # s = int(''.join(s[::-1]))
    # print(s)

File: leetcode/reverse/reverse.cpp
#include <iostream>
#include <limits.h>
using namespace std;

// https://www.cnblogs.com/grandyang/p/4125588.html
class Solution {
public:
    int reverse(int x) {
        int res = 0;
        while (x != 0) {
            if (abs(res) > INT_MAX / 10) return 0;
            res = res * 10 + x % 10;
            x /= 10;
        }
        return res;
    }
};


int main(){
    int x = 123;
    // int sign = x > 0 ? 1 : -1;
    // cout << sign << endl;
    // string s = to_string(x);
    // cout << s << endl;
    // reverse(s.begin(),s.end());
    // cout << s << endl;
    // cout << stoi(s) << endl;
    Solution solver;
    cout << solver.reverse(123) << endl;
    return 0;
}